src/core/linux-aio.cc

```diff
@@ -171,9 +171,9 @@ void setup_aio_context(size_t nr, linux_abi::aio_context_t* io_context) {
     auto r = io_setup(nr, io_context);
     if (r < 0) {
         char buf[1024];
-        char *msg = strerror_r(errno, buf, sizeof(buf));
+        strerror_r(errno, buf, sizeof(buf));
```

Comment:
this could be wrong. as per the manapge

>        The  GNU-specific  strerror_r()  returns  a pointer to a string containing the error message.  This may be either a pointer to a string that the function stores in buf, or a pointer to some (immutable) static string (in which
>        case buf is unused).  If the function stores a string in buf, then at most buflen bytes are stored (the string may be truncated if buflen is too small and errnum is unknown).  The string always  includes  a  terminating  null
>        byte ('\0').

actually, glibc does not use the `buf` passed to it if `_GNU_SOURCE` is defined. see also https://godbolt.org/z/nv8qzMK8K

probably we can check for the existence of glibc's extension by checking the `_GNU_SOURCE` macro
   ```c++
   char buf[1024] = "unknown error";
   #ifdef _GNU_SOURCE
       char *msg = strerror_r(error, buf, 0);
   #else
       char *msg = strerror_r(error, buf, sizeof(buf));
   #endif
   ```

also, i'd suggest put more details in the commit message to explain that what we are using now is a glibc extension of `strerror_r()`, which does not exists in, for instance, musl libc.

Reply:
okay thanks, will change it

Reply:
Tried that and it looks like alpine has _GNU_SOURCE defined somewhere or maybe in boost or another header?

```
src/seastar-1ddb6fe061aea521aeeafe3ad6ab926b803ab51d/src/core/linux-aio.cc:175:31: error: invalid conversion from 'int' to 'char*' [-fpermissive]
  175 |         char *msg = strerror_r(errno, buf, 0);
      |                     ~~~~~~~~~~^~~~~~~~~~~~~~~
      |                               |
      |                               int
```

Also the musl version just returns and `int` https://git.musl-libc.org/cgit/musl/tree/src/string/strerror_r.c

This works but needs `features.h` to get `__GLIBC__`
```
#ifdef __GLIBC__
        char *msg = strerror_r(errno, buf, sizeof(buf));
 #else
        char *msg = buf;
       strerror_r(errno, buf, sizeof(buf));
 #endif
```

Another solution is to implement with `snprintf` http://www.club.cc.cmu.edu/~cmccabe/blog_strerror.html

Reply:
> Tried that and it looks like alpine has _GNU_SOURCE defined somewhere or maybe in boost or another header?
> 
> ```
> src/seastar-1ddb6fe061aea521aeeafe3ad6ab926b803ab51d/src/core/linux-aio.cc:175:31: error: invalid conversion from 'int' to 'char*' [-fpermissive]
>   175 |         char *msg = strerror_r(errno, buf, 0);
>       |                     ~~~~~~~~~~^~~~~~~~~~~~~~~
>       |                               |
>       |                               int
> ```
> 
> Also the musl version just returns and `int` https://git.musl-libc.org/cgit/musl/tree/src/string/strerror_r.c
> 
> This works but needs `features.h` to get `__GLIBC__`

how about: 
```c++
char buf[1024];
#if (_POSIX_C_SOURCE >= 200112L) && !  _GNU_SOURCE
    const char *msg = strerror_r(error, buf, sizeof(buf)) ? "unknown error" : buf;
#else
    const char * msg = strerror_r(error, buf, 0);
#endif

Reply:
the musl/posix version returns an int https://linux.die.net/man/3/strerror_r

Reply:
> the musl/posix version returns an int https://linux.die.net/man/3/strerror_r

i understand. and the first branch does check the `int` return value. am i missing anything?

Reply:
yes, you suggestion has both cases with `const char *msg = strerror_r`

which fails on musl as it returns an int

Reply:
@a16bitsysop please reread the statement:
```c++
const char *msg = strerror_r(error, buf, sizeof(buf)) ? "unknown error" : buf;
```
which checks the return value of `strerror_r()`, if it returns non-zero, "unknown error" is returned, `buf` otherwise.

Reply:
ah, I see.  The `_GNU_SOURCE` check wasn’t working as its already set by one of the dependency headers I think.

Reply:
> ah, I see. The `_GNU_SOURCE` check wasn’t working as its already set by one of the dependency headers I think.

hmm, i tested locally last week end. will retry later this week.

Reply:
okay. just tested. my approach indeed does work. =(